include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -103,6 +107,197 @@ void FFT(NBL_REF_ARG(Accessor) sharedmemAccessor, NBL_REF_ARG(complex_t<Scalar>)
     }
 }
 
+*/
+
+// ----------------------------------------- ABOVE MARKED FOR DELETION -----------------------------------------------------
+
+template<uint16_t K, bool Inverse, typneame Scalar, class device_capabilities=void>
+struct FFT;
+
+// For the FFT methods below, we assume:
+//      - Accessor is a global memory accessor to an array fitting 2 * _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar>, used to get inputs / set outputs of the FFT
+//      - SharedMemoryAccessor accesses a shared memory array that can fit _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar> 
+
+// 2 items per invocation forward specialization
+template<typename Scalar, class device_capabilities>
+struct FFT<2,false,device_capabilities>
+{
+    template<typename SharedMemoryAccessor>
+    static void FFT_loop(uint32_t stride, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi, uint32_t threadID, NBL_REF_ARG(MemoryAdaptor<SharedMemoryAccessor>) sharedmemAdaptor)
+    {
+        const bool topHalf = (threadID & stride) != 0;
+        vector <Scalar, 2> toTrade = topHalf ? vector <Scalar, 2>(lo.real(), lo.imag()) : vector <Scalar, 2>(hi.real(), hi.imag());
+        
+        // Block memory writes until all threads are done with previous work
+        sharedmemAdaptor.workgroupExecutionAndMemoryBarrier();
+        sharedmemAdaptor.set(threadID, toTrade);
+        
+        // Wait until all writes are done before reading
+        sharedmemAdaptor.workgroupExecutionAndMemoryBarrier();
+        sharedmemAdaptor.get(threadID ^ stride, toTrade);
+        
+        if (topHalf)
+        {
+            lo.real(toTrade.x);
+            lo.imag(toTrade.y);
+        }
+        else
+        {
+            hi.real(toTrade.x);
+            hi.imag(toTrade.y);
+        }
+        // Get twiddle with k = threadID mod stride, N = 2 * stride
+        fft::DIF<Scalar>::radix2(fft::twiddle<false, Scalar>(threadID & (stride - 1), stride << 1), lo, hi);    
+    }
+
+
+    template<typename Accessor, typename SharedMemoryAccessor>
+    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
+    {
+        // Set up the MemAdaptors
+        MemoryAdaptor<Accessor> memAdaptor;
+        memAdaptor.accessor = accessor;
+        MemoryAdaptor<SharedMemoryAccessor> sharedmemAdaptor;
+        sharedmemAdaptor.accessor = sharedmemAccessor;
+
+        // Compute the SubgroupContiguousIndex only once
+        const uint32_t threadID = uint32_t(SubgroupContiguousIndex());
+
+        // Read lo, hi values from global memory
+        vector <Scalar, 4> loHiPacked;
+        memAdaptor.get(threadID, loHiPacked);
+        complex_t<Scalar> lo = {loHiPacked.x, loHiPacked.y};  
+        complex_t<Scalar> hi = {loHiPacked.z, loHiPacked.w};
+
+        // special first iteration
+        fft::DIF<Scalar>::radix2(fft::twiddle<false, Scalar>(threadID, _NBL_HLSL_WORKGROUP_SIZE_ << 1), lo, hi); 
```

Comment:
you need to predicate this on `_NBL_HLSL_WORKGROUP_SIZE>SubgroupSize` otherwise you end up doing the special step twice when your workgroup-size == subgroup size

(you don't need to worry about WorkgroupSize < SubgroupSize because that goes against `comptueFullSubgroups` pipeline creation option that you'll require)